mountpoint-s3-crt/src/s3/endpoint_resolver.rs

```diff
@@ -0,0 +1,265 @@
+                self.inner.as_ptr(),
+                name.as_aws_byte_cursor(),
+                value.as_aws_byte_cursor(),
+            )
+            .ok_or_last_error()
+        }
+    }
+
+    /// Add the parameter to request context whose value is in form of boolean
+    pub fn add_boolean(&mut self, allocator: &Allocator, name: &OsStr, value: bool) -> Result<(), Error> {
+        // SAFETY: allocator.inner and self.inner should be valid pointers.
+        // `name` should be valid aws byte cursor not be modified further.
+        unsafe {
+            aws_endpoints_request_context_add_boolean(
+                allocator.inner.as_ptr(),
+                self.inner.as_ptr(),
+                name.as_aws_byte_cursor(),
+                value,
+            )
+            .ok_or_last_error()
+        }
+    }
+}
+
+impl Drop for RequestContext {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` is a valid `aws_endpoints_request_context`, and on Drop it's safe to decrement
+        // the reference count since we won't use it again through `self.`
+        unsafe {
+            aws_endpoints_request_context_release(self.inner.as_ptr());
+        }
+    }
+}
+
+/// Resolved endpoint for the given context using rule engine and rule set
+#[derive(Debug)]
+pub struct ResolvedEndpoint {
+    inner: NonNull<aws_endpoints_resolved_endpoint>,
+}
+
+impl ResolvedEndpoint {
+    /// Get URI from the Resolved Endpoint
+    pub fn get_url(&self, allocator: &mut Allocator) -> Result<Uri, Error> {
+        let mut url: aws_byte_cursor = Default::default();
+        // SAFETY: self.inner is valid pointer and url is passed as valid mutable pointer.
+        //`aws_endpoint_resolved_enpoint_get_url` ensures to return an initialized aws_byte_cursor for url or else it will return an error.
+        unsafe {
+            aws_endpoints_resolved_endpoint_get_url(self.inner.as_ptr(), &mut url).ok_or_last_error()?;
+            let uri = aws_byte_cursor_as_slice(&url);
+            let uri = OsStr::from_bytes(uri);
+            Uri::new_from_str(allocator, uri)
+        }
+    }
+}
+
+impl Drop for ResolvedEndpoint {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` is a valid `aws_endpoints_resolved_endpoint`, and on Drop it's safe to decrement
+        // the reference count since we won't use it again through `self.`
+        unsafe {
+            aws_endpoints_resolved_endpoint_release(self.inner.as_ptr());
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::ffi::OsStr;
+
+    use crate::common::allocator::Allocator;
+
+    use super::{RequestContext, RuleEngine};
+
+    #[test]
+    fn test_regions_outside_aws_partition() {
+        let mut new_allocator = Allocator::default();
+        let endpoint_rule_engine = RuleEngine::new(&new_allocator).expect("Rule Engine not constructed");
+        let mut endpoint_request_context =
+            RequestContext::new(&new_allocator).expect("Request Context not constructed");
```

Comment:
expect is good to have, so long as the message is meaningful.

let's use `unwrap()` here and later for endpoint_uri. the expect messages for `resolve` are good!